src/ucp/core/ucp_request.c

```diff
@@ -411,18 +411,19 @@ ucp_request_send_start(ucp_request_t *req, ssize_t max_short,
 
 void ucp_request_send_state_ff(ucp_request_t *req, ucs_status_t status)
 {
-    /*
-     * FIXME should not fast-forward requests owned by UCT
-     */
     ucp_trace_req(req, "fast-forward with status %s", ucs_status_string(status));
 
     if (req->send.state.uct_comp.func == ucp_ep_flush_completion) {
         ucp_ep_flush_request_ff(req, status);
-    } else if (req->send.state.uct_comp.func) {
-        req->send.state.dt.offset      = req->send.length;
-        req->send.state.uct_comp.count = 0;
+    } else if (req->send.state.uct_comp.func != NULL) {
         uct_completion_update_status(&req->send.state.uct_comp, status);
-        req->send.state.uct_comp.func(&req->send.state.uct_comp);
+
+        /* If there are no in-progress operations, complete the operation,
+         * otherwise - just update status and wait for completion of operations
+         */
+        if (req->send.state.uct_comp.count == 0) {
```

Comment:
does it mean that failed requests will wait for completion too?

Reply:
yes, transport will provide a completion too in case of error